File: in_Java/0105 Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer,Integer> hm = new HashMap<>();
        for(int i = 0; i < inorder.length; i++){
            hm.put(inorder[i],i);
        }
        TreeNode root = helper(preorder, 0, preorder.length-1, inorder, 0, inorder.length-1, hm);
        return root;
    }
    private TreeNode helper(int[] pre, int preStart, int preEnd, int[] in, int inStart, int inEnd, HashMap<Integer,Integer> hm){
        if(preStart > preEnd || inStart > inEnd) return null;
        TreeNode root = new TreeNode(pre[preStart]);
        int id = hm.get(root.val);
        int dis = id - inStart;
        root.left = helper(pre, preStart+1, preStart + dis, in, inStart, id-1, hm);
        root.right = helper(pre, preStart+dis+1, preEnd, in, id+1, inEnd, hm);
        return root;
    }
}

File: in_Python/0105 Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if inorder:
            r = preorder.pop(0)
            idx = -1
            for i,x in enumerate(inorder):
                if x == r:
                    idx = i
                    break
            root = TreeNode(r)
            root.left = self.buildTree(preorder, inorder[:idx])
            root.right = self.buildTree(preorder, inorder[idx+1:])
            return root